### PR TITLE
Ruby Capstone group project: Catalog of my things

### PR DESCRIPTION
**Pull request by Piercel2022**
In this project, we create a console app that will help you to keep a record of different types of things you own: books, music albums, movies, and games. Everything will be based on the UML class diagram presented below. The data will be stored in JSON files but you will also prepare a database with tables structure analogical to your program's class structure.
catalog_of_my_things

Video link - https://drive.google.com/file/d/1NZ3us7FJ7ZUKsc1TWcfK_Z7A6Okj7rvl/view?usp=sharing

### CODE REVIEW

File: classes/genre.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+class Genre
+  def initialize(name, id = Random.rand(1...1000))
+    @id = id
+    @name = name
+    @items = []
+  end
+
+  def add_item(album)
+    @items << album
+  end
```
Comment (mirouhml) [inline classes/genre.rb]: - [x] Please know that you need to specify the correct 1-to-many-relationship here, this is an example of how it should be: 
```ruby
    def add_item(item)
      @items.push(item)
      item.genre = self
    end
```

File: classes/author.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(first_name, last_name, id = Random.rand(1..1000))
+    @id = id
+    @first_name = first_name
+    @last_name = last_name
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (mirouhml) [inline classes/author.rb]: - [x] Please know that you need to specify the correct 1-to-many-relationship here, this is an example of how it should be: 
```ruby
    def add_item(item)
      @items.push(item)
      item.genre = self
    end
```

File: classes/label.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(title, color, id = Random.rand(1..1000))
+    @id = id
+    @title = title
+    @color = color
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (mirouhml) [inline classes/label.rb]: - [x] Please know that you need to specify the correct 1-to-many-relationship here, this is an example of how it should be: 
```ruby
    def add_item(item)
      @items.push(item)
      item.genre = self
    end
```

File: classes/source.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Source
+  def initialize(name, id = Random.rand(1..1000))
+    @id = id
+    @name = name
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (mirouhml) [inline classes/source.rb]: - [x] Please know that you need to specify the correct 1-to-many-relationship here, this is an example of how it should be: 
```ruby
    def add_item(item)
      @items.push(item)
      item.genre = self
    end
```

File: classes/item.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  def add_label(label)
+    @label = label
+    label.add_item(self)
+  end
+
+  def add_genre(genre)
+    @genre = genre
+    genre.add_item(self)
+  end
```
Comment (mirouhml) [inline classes/item.rb]: - [x] Please know that the Item class should have one Genre, Author, Label, and Source, to make sure that's the case you can do it like this:

```ruby
    def genre=(genre)
      @genre = genre
      genre.items.push(self) unless genre.items.include?(self)
    end

    def author=(author)
      @author = author
      author.items.push(self) unless author.items.include?(self)
    end

    def label=(label)
      @label = label
      label.items.push(self) unless label.items.include?(self)
    end
    
    def source=(source)
      @source = source
      source.items.push(self) unless source.items.include?(self)
    end
```

File: classes/item.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  # 03 Implement methods:
+  # can_be_archived?() in the Item class
+  # should return true if published_date is older than 10 years
+  # otherwise, it should return false
+  def can_be_archived?
+    true if (Date.today.to_time.year - @date_published.to_time.year) > 10
+  end
```
Comment (mirouhml) [inline classes/item.rb]: - [x] Please know that this method should return false when the condition isn't met, and easy way to do this on ruby is this:

```ruby
    def can_be_archived?
      today = Date.today
      10_years_ago = Date.new(today.year - 10, today.month, today.day)

      @published_date < 10_years_ago
    end
```

File: schema.sql [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+  on_spotify BOOLEAN,
+  publish_date DATE,
+  archived BOOLEAN
+);
+
+CREATE TABLE genre (
+  id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+  name VARCHAR(200)
+);
```
Comment (mirouhml) [inline schema.sql]: - [x] Please know that you should set up foreign keys correctly for the 4 main tables, music_album, game, movie and book, according to the UML diagram they are all linked to genre, label, author and source, this means that they should have foreign keys for all of them, kindly refer back to the UML diagram and add what's missing:
image

**Review by mirouhml — CHANGES_REQUESTED** [review]
# REQUIRED CHANGES 🔴

Hi @Piercel2022 @ishpaul777 @29td @PNdunguMaina, 

Good job so far!

There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- Great job setting up the correct linters ✔️
- Highly organized and clean code ✔️ 👨🏻‍💻
- Professional documentation ✔️
- Correct workflow ✔️
- Meaningful commit messages ✔️

## Required Changes ♻️
_Check the comments under the review._

- [x] Please know that you should have recorded a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

  - A description of the project.
  - A demo of the project features.
  - You should also highlight some interesting piece of code or something you built that you are very proud of.
  - You all should appear in the video and talk about the project.

  For recording the video you can use the Zoom recording features while in a call with your peers.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
# STATUS: APPROVED :green_circle: :clap: 

Hello  @Team  , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

HIGHLIGHTS:

- Your project is professional 
- Your repo is professional 
- No Linter Errors
- Unit tests are passing




Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 

well-1

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**